README

	SubC Compiler, Version 2015-02-02
	By Nils M Holm, 2011--2015
	Placed in the public domain


	SUMMARY

	SubC is a compiler for a (mostly) strict and sane subset of
	C as described in "The C Programming Language", 2nd Ed (also
	known informally as "ANSI C" or "C89").

	A previous version of the compiler is described in great detail
	in the book "Practical Compiler Construction", which can be
	purchased at Lulu.com. See  http://www.t3x.org/reload/  for
	ordering information.

	The SubC compiler can compile itself. Unlike many other small C
	compilers, it does not bend the rules, though. Its code passes
	"gcc -Wall -pedantic" with little or no warnings (depending on
	the gcc version used). Of course, you can also bootstrap it with
	other C compilers, such as Clang or PCC.

	SubC is fast and simple. Its output is statically linked (where
	available) and typically small due to a non-bloated library). It
	uses a simple optimizer on per-expression basis.


	SUPPORTED SYSTEMS

	SubC generates code for GAS, the GNU assembler (except for the
	DOS version, which emits TASM-style syntax). It targets the
	following processors and operating systems:

		FreeBSD		386	armv6	x86-64
		Linux		386	-	x86-64
		NetBSD		-	-	x86-64
		OpenBSD		386%	-	-
		Windows/MinGW	386	-	-
		Darwin		-	-	x86-64*%
		DOS		8086!

		  %  uses the syscall layer of the host libc
		  *  untested
		  !  experimental
		 (/) broken

	Platforms tagged "untested" are not regularly tested by myself
	and are therefore subject to potential bit rot. You can help
	me improve SubC by running "make tests" on an "untested"
	platform and let me know about the results.

	Platforms using the system's libc as a thin system call layer
	often cause build/stability problems due to the omnipresence of
	the GNU libc, which is not "thin" at all. Expect trouble on
	those systems!

	Platforms tagged "broken" currently will not compile or run
	properly for some reason. See the Todo file for details.

	The DOS version brings its own toolchain, which can be found in
	the s86/ directory, so no pre-existing DOS assembler or linker
	is required to compile SubC programs on DOS.

	Porting SubC to other 32-bit or 64-bit platforms should be
	quite straight-forward. See the file "Porting" and/or the book
	for a general road map.


	CHANGES TO THE BOOK VERSION

	Note: The book version runs on FreeBSD/386 exclusively.

	The current version uses an improved code generator, which
	emits much smaller and faster code than the book compiler.
	The techniques are described in the book, though.

	The current version of the SubC compiler adds support for
	the following parts of C language to the version described
	in "Practical Compiler Construction":

	o  &array is now valid syntax (you no longer have to write
	   &array[0]).

	o  the auto, register, and volatile keywords are recognized
	   (as no-ops). Yes, volatile is safe, because SubC does not
	   have register variables.

	o  enums may now be local.

	o  extern identifiers may now be declared locally.

	o  Prototypes may have the static storage class.

	o  There is support for structs and unions.

	o  jmp_buf is now a struct; setjmp() and longjmp() must be
	   called with &jmp_buf.

	o  FILEs are now structs and can no longer be mistaken for
	   ints by the type checker.

	o  The #error, #line, and #pragma commands have been added.

	o  There is a (non-standard) kprintf() function, which is
	   like fprintf(), but uses a file descriptor.

	o  There is now a (slightly incompatible) varargs mechanism.
	   Here is how it works:

		#include <varargs.h>

		void p(int a, int b, ...) {
			int	first;
			void	*ap;

			ap = _va_start(&b);
			first = (int) _va_arg(&ap);
			vprintf("other args: %d %d %d\n", ap);
			_va_end(&ap);
		}

	o  The vprintf(), vfprintf(), and vsprintf() functions have
	   been added to the runtime library.

	o  A broader subset of C expression syntax is accepted
	   in constant expression contexts. For example, pointer
	   variables can be initialized with NULL.


	DIFFERENCES BETWEEN SUBC (THIS VERSION) AND FULL C89

	o  The following keywords are not recognized:
	   const, double, float, goto, long, short, signed, typedef,
	   unsigned.

	o  There are only two primitive data types: the signed int and
	   the unsigned char; there are also void pointers, and there
	   is limited support for int(*)() (pointers to functions
	   of type int).

	o  No more than two levels of indirection are supported, and
	   arrays are limited to one dimension, i.e. valid declarators
	   are limited to x, x[], *x, *x[], **x (and (*x)()).

	o  K&R-style function declarations (with parameter declarations
	   between the parameter list and function body) are not
	   accepted.

	o  There are no ``const'' variables.

	o  There is no typedef.

	o  There are no unsigned integers, long integers, or signed
	   chars.

	o  Struct/union declarations must be separate from the
	   declarations of struct/union objects, i.e.
	   ``struct p { int x, y; } q;'' will not work.

	o  Struct/union declarations must be global (struct and union
	   objects may be declared locally, though).

	o  There is no support for bit fields.

	o  Only ints, chars, and arrays of int and char can be
	   initialized in their declarations; pointers can be
	   initialized with 0 or NULL.

	o  Local arrays cannot have initializers.

	o  Local declarations are limited to the beginnings of function
	   bodies (they do not work in other compound statements).

	o  Arguments of prototypes must be named.

	o  There is no goto.

	o  There are no parameterized macros.

	o  The #if and #elif preprocessor commands are not recognized.

	o  The preprocessor does not accept multi-line commands.

	o  The preprocessor does not accept comments in (some) commands.

	o  The preprocessor does not recognize the # and ## operators.

	o  There may not be any blanks between the # that introduces
	   a preprocessor command and the subsequent command (e.g.:
	   "# define" would not be recognized as a valid command).

	o  The sizeof operator requires parentheses.

	o  Subscripting an integer with a pointer (e.g. 1["foo"]) is
	   not supported.

	o  Function pointers are limited to one single type, int(*)(),
	   and they have no argument types.

	o  There is no assert() due to the lack of parameterized macros.

	o  The atexit() mechanism is limited to one function (this may
	   even be covered by TCPL2).

	o  The signal() function returns int due to the lack of a more
	   sophisticated type system; the return value must be casted to
	   int(*)() manually.

	o  Most of the time-related functions are missing, in particular:
	   asctime(), gmtime(), localtime(), mktime(), and strftime().

	o  The clock() function is missing, because CLOCKS_PER_SEC
	   varies among systems.

	o  The ctime() function ignores the time zone.

	o  The varargs mechanism is slightly incompatible.


	SELECTING A TARGET PLATFORM

	The easiest way to prepare a build is to run the configure
	script in this directory. Don't worry, it is just a simple
	script that will figure out the host platform via uname and
	link a few machine-dependent files into place.

	If you want to configure the compiler manually: select one of
	the target descriptions (cg*.c) files in src/targets/cg and
	symlink it to src/cg.c. Also link the corresponding header
	file into place:

		(cd src && ln -fs targets/cg/cg386.c cg.c)
		(cd src && ln -fs targets/cg/cg386.h cg.h)

	Next select the C startup (crt0) file for your OS and CPU type
	from src/targets/OS-CPU/ and link it to src/lib/crt0.s, e.g.:

		(cd src/lib && \
		 ln -fs ../targets/freebsd-386/crt0-freebsd-386.s \
		 	crt0.s)

	If your OS/CPU combination is not supported, you might try
	to port the compiler. See the file "Porting" for details.

	You will also need some operating system-dependent
	definitions, which are kept in files named "sys-OS-CPU.h"
	in src/targets/OS-CPU/. Just symlink the appropriate file
	to src/sys.h:

		(cd src && \
		 ln -fs targets/freebsd-386/sys-freebsd-386.h sys.h)


Changes
2015-02-02 Fix: Long init lists in array declarations could run out of nodes.
2014-12-07 Fix: strncpy() should pad destination. Thanks, Alexey Frunze!
2014-12-07 Fix: fseek() should return 0 on success. (also by Alexey)
2014-12-07 Fix: longjmp() may not pass 0 to setjmp(). (also by Alexey)
2014-12-07 Fixed emitted byte shift code on x86. Thanks, Teran McKinney!
2014-06-22 Running extended triple test only in case of failure.
2014-06-03 Fix: syntax errors in constexprs could crash the compiler.
2014-05-25 Updated and re-organized documentation.
2014-05-24 Added operation reordering (tree skewing), e.g.: a+b*c --> b*c+a.
2014-05-24 Added strength reduction 2/2^N --> 2>>N.
2014-05-24 Added strength reduction 2*2^N --> 2<<N.
2014-05-20 Fix: convert '/' in all commands paths in Windows system().
2014-05-19 Applied patch for fixing system() on Windows. Thanks, Jean-Marc!
2014-05-18 Added OpenBSD ELF note.
2014-05-18 Fix: missed some functions when adapting OpenBSD crt0 to R-L calls.
2014-05-18 Applied more Windows patches by Jean-Marc Lienher. Thanks!
2014-05-17 Updated documentation.
2014-05-17 Changed random number test in libtest.c.
2014-05-17 Made kprintf() convert '\n' on descriptors 1,2 on DOS.
2014-05-16 Made some stdio functions emit CR,LF for '\n' on stdout/err on DOS.
2014-05-14 Added triple test script to gather data in case of failure.
2014-05-14 Added some DOS #ifdef:s to libtest.c (./file vs .\file).
2014-05-14 Adapted printf test in libtest.c for 16-bit systems.
2014-05-13 Fixed setjmp() function in DOS port (overwrote caller's memory).
2014-05-12 Updated documentation.
2014-05-10 Fix: added installation of (non-stdio) ncrt0.o to Makefile.
2014-05-10 Applied a load of Windows patches by Jean-Marc Lienher. Thanks!
2014-05-09 Fix: wrong Opt_sum_lim when BPW != sizeof(int).
2014-05-09 Removed constant expr parser, using constant folding instead.
2014-05-05 Increased NODEPOOLSZ.
2014-05-05 Windows port builds, but programs crash; back to EXPERIMENTAL.
2014-05-05 Changed winbuild.bat to print instructions, using make again.
2014-05-04 Changed triple test to use assembly output instead of binaries.
2014-05-04 Added -N (don't use stdio) flag and generation of ncrt0's.
2014-05-03 Removed "make" from winbuild.bat script (probably still broken).
2014-05-03 Fix: remove trailing '\r' in preprocessor lines.
2014-05-03 Gave up writing a signal(3) function for NetBSD. Using libc for now.
2014-05-03 Removed almost all libc dependencies from NetBSD-x86-64 crt0.
2014-05-02 Removed libc dependencies from FreeBSD-x86-64 crt0.
2014-05-01 Fixed order of argc/argv on DOS port (due to changed calling conv).
2014-05-01 Marked all untested ports "untested".
2014-05-01 Added removal of zero offsets from address computation.
2014-04-30 Adapted all remaining ports to new calling convention.
2014-04-29 Adapted NetBSD-x86-64 crt0 to new calling convention.
2014-04-29 Adapted FreeBSD-x86-64 crt0, build failed, marked BROKEN.
2014-04-28 Adapted FreeBSD-armv6 crt0 to new calling convention.
2014-04-28 Adapted FreeBSD-386 crt0 to new calling convention.
2014-04-28 Removed varargs redefinition hack (no longer needed).
2014-04-28 Changed calling convention to standard C (right-to-left).
2014-04-27 Added optimizer.
2014-04-26 Added tree abstraction.
2014-04-20 Forked this version.
2014-04-07 Fixed XSTRIP and SNM programs on 16-bit systems.
2014-04-05 Made s86 cache code for faster resolution of forward references.
2014-04-05 Fix: missing file name in error messages due to refactoring.
2014-04-03 Added DOS build script (targets/dos-8086/dosbuild.bat).
2014-04-02 Clean-up and misc. small fixes.
2014-04-01 The compiler can now be cross-compiled to DOS.
2014-04-01 Updated codumentation.
2014-04-01 Added DOS/EXE targets to Makefiles.
2014-03-31 Made compiler controller a separate executable on DOS.
2014-03-31 Factored out compiler controller.
2014-03-29 Documented bootstrapping of the DOS version.
2014-03-28 Added minimal runtime for compiling SCC86 on DOS.

s86/sld.c
/*
 *	S86 -- A Simple 8086 Tool Chain
 *	EXE Loader
 *	Nils M Holm, 1993,1995,2013,2014
 *	In the public domain
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "defs.h"
#include "arc.h"
#include "sym.h"
#include "obj.h"
#include "exe.h"

/*
 * Resolve dependencies in memory;
 * fast, but will not work on 16-bit systems.
 */
/* #define INCORE	/**/

#define LDCODE	"ldc00000.tmp"
#define LDDATA	"ldd00000.tmp"
#define LDRLOC	"ldr00000.tmp"
#define LDRBUF	"ldb00000.tmp"

#define SYM_SZ	20480
#define EXT_SZ	16384
#define BUF_SZ	60000

int	debug = 0;

FILE	*Objf, *Ctmp, *Dtmp, *Rtmp, *Rbuf, *Exef;
char	Outfile[81];
int	Errors;
char	Ohd[OHDSZ];
char	Xhd[XHDR_SZ];
int	Codep, Datap;
int	Stkinit;

char	Buf[BUFSIZ];

char	DOSRel[NDOSREL*4];

char	*Libdir;

char	*Buffer;
char	*Symtab;	/* Shared extrn/public */
char	*Extsym;
char	*Reloc;
int	N_dosrel;

char	Arh[ARH_LEN];

#define LOGFILE	stdout

void error(char *m, char *n) {
	if (n)
		printf("sld: %s: %s\n", m, n);
	else
		printf("sld: %s\n", m);
	Errors++;
}

void fatal(char *m) {
	printf("sld: fatal error: ");
	error(m, NULL);
	if (!debug) {
		remove(LDCODE);
		remove(LDDATA);
		remove(LDRLOC);
		remove(LDRBUF);
	}
	exit(1);
}

void readerr(void) {
	fatal("file read error");
}

void writeerr(void) {
	fatal("file write error");
}

void init(void) {
	int	i;
	char	*nomem;
	char	*notmp;

	nomem = "not enough memory";
	notmp = "could not create temp file: ld?00000";
	if ((Symtab = malloc(SYM_SZ)) == NULL)
		fatal(nomem);
	if ((Extsym = malloc(EXT_SZ)) == NULL)
		fatal(nomem);
	memset(Symtab, 0, SYM_SZ);
	memset(Extsym, 0, EXT_SZ);
#ifdef INCORE
	if ((Buffer = malloc(BUF_SZ)) == NULL)
		fatal(nomem);
	memset(Buffer, 0, BUF_SZ);
#endif
	if ((Ctmp = fopen(LDCODE, "wb+")) == NULL)
		fatal(notmp);
	if ((Dtmp = fopen(LDDATA, "wb+")) == NULL) {
		remove(LDCODE);
		fatal(notmp);
	}
	if ((Rtmp = fopen(LDRLOC, "wb+")) == NULL) {
		remove(LDCODE);
		remove(LDDATA);
		fatal(notmp);
	}
#ifndef INCORE
	if ((Rbuf = fopen(LDRBUF, "wb+")) == NULL) {
		remove(LDCODE);
		remove(LDDATA);
		remove(LDRLOC);
		fatal(notmp);
	}
#endif
	if ((Exef = fopen(Outfile, "wb+")) == NULL) {
		error("could not create output file", Outfile);
		fatal("aborting");
	}
	for (i = 0; i<SYM_SZ; i += SSIZE)
		Symtab[i] = 0;
	for (i=0; i < EXT_SZ-MKSZ; i += MKSZ)
		Extsym[i+MKADDR] = Extsym[i+MKADDR+1] = 0;
	N_dosrel = 0;
	Codep = Datap = 0;
	Stkinit = 1024;
}

int addsym(char *buf) {
	int	i, add, pub;

	pub = (buf[SCLASS]==PUBLIC);
	for (add=-1,i=0; i<SYM_SZ; i+=SSIZE) {
		if (!Symtab[i]) {
			if (add != -1) continue;
			memcpy(&Symtab[i], buf, SSIZE);
			Symtab[i+SMARK] = Symtab[i+SMARK+1] = 0xff;
			if (debug)
				fprintf(LOGFILE,
					"add sym:\t%s (%c,%c) = %04x\n",
					buf, buf[SSEGMT], buf[SCLASS],
					buf[SADDR] + (buf[SADDR+1] << 8));
			add = i;
		}
		else if (pub) {
			if (	!strcmp(&Symtab[i], buf) &&
				Symtab[i+SCLASS] == PUBLIC
			)
				error("duplicate symbol", buf);
		}
	}
	if (add != -1) return add;
	fatal("symbol table overflow");
	return -1;
}

void mkmark(int si, char *ebuf) {
	int	a, i;

	for (i=0; i<EXT_SZ; i+= MKSZ) {
		if (Extsym[i+MKADDR] == 0 && Extsym[i+MKADDR+1] == 0) {
			a = Symtab[si+SMARK] + (Symtab[si+SMARK+1] << 8);
			Extsym[i+MKPTR] = a;
			Extsym[i+MKPTR+1] = a>>8;
			Symtab[si+SMARK] = i;
			Symtab[si+SMARK+1] = (i >> 8);
			Extsym[i+MKFLAG] = ebuf[MKFLAG];
			Extsym[i+MKADDR] = ebuf[MKADDR];
			Extsym[i+MKADDR+1] = ebuf[MKADDR+1];
			if (debug)
				fprintf(LOGFILE,
					"mark:\t%s\tF=%02x\tA=%04x\tP=%04x\n",
					&Symtab[si], Extsym[i+MKFLAG],
					Extsym[i+MKADDR] +
						(Extsym[i+MKADDR+1]<<8),
					Extsym[i+MKPTR] +
						(Extsym[i+MKPTR+1]<<8));
			return;
		}
	}
	fatal("mark table overflow");
}

void load(char *name) {
	int	i, n, off, si, nrel, old, cmp, size, t, k, a;
	char	*paddr;
	char	sym[SSIZE];
	char	ee[MKSZ];
	char	drel[2];

	if (fread(Ohd, 1, OHDSZ, Objf) != OHDSZ) {
		error("bad object file (short file)", name);
		return;
	}
	if (Ohd[OMAGIC] != O_MAGIC) {
		error("bad object file (magic match failed)", name);
		return;
	}
	off = OHDSZ;
	while (off < Ohd[OEXTP] + (Ohd[OEXTP+1]<<8)) {
		if (fread(sym, 1, SSIZE, Objf) != SSIZE)
			readerr();
		paddr = &sym[SADDR];
		k = paddr[0] + (paddr[1]<<8);
		k += sym[SSEGMT] == SCODE? Codep: Datap;
		paddr[0] = k;
		paddr[1] = k>>8;
		addsym(sym);
		off += SSIZE;
	}
	while (off < Ohd[ODOSREL] + (Ohd[ODOSREL+1]<<8)) {
		if (fread(sym, 1, SSIZE, Objf) != SSIZE)
			readerr();
		si = addsym(sym);
		off += SSIZE;
		fread(ee, 1, MKSZ, Objf);
		off += MKSZ;
		while (ee[MKADDR] != 0xff || ee[MKADDR+1] != 0xff) {
			k = ee[MKADDR] + (ee[MKADDR+1]<<8);
			k += Codep;
			ee[MKADDR] = k;
			ee[MKADDR+1] = k>>8;
			mkmark(si, ee);
			fread(ee, 1, MKSZ, Objf);
			off += MKSZ;
		}
	}
	while (off < Ohd[ORELOC] + (Ohd[ORELOC+1]<<8)) {
		if (fread(drel, 1, 2, Objf) != 2) readerr();
		off += 2;
		n = Codep + drel[0] + (drel[1]<<8);
		if (debug) fprintf(LOGFILE, "DOS reloc:\t%04x\n", n);
		drel[0] = n;
		drel[1] = n>>8;
		if (fwrite(drel, 1, 2, Exef) != 2) writeerr();
		drel[0] = drel[1] = 0;
		if (fwrite(drel, 1, 2, Exef) != 2) writeerr();
		N_dosrel++;
	}
	n = Ohd[OCODE] + (Ohd[OCODE+1]<<8) - off;
	off += n;
	nrel = n/3;
	rewind(Rtmp);
	for (i=n; i>0; ) {
		k = (char *) i < (char *) BUFSIZ? i: BUFSIZ;
		if (fread(Buf, 1, k, Objf) != k) readerr();
		if (fwrite(Buf, 1, k, Rtmp) != k) writeerr();
		i -= k;
	}
	if (debug) fprintf(LOGFILE, "%d RELOC entries\n", nrel);
	n = Ohd[ODATA] + (Ohd[ODATA+1]<<8) - off;
	off += n;
	if ((char *) n >= (char *) CODESZ) fatal("text area too large");
	if (debug) {
		fprintf(LOGFILE, "%d TEXT bytes\n", n);
		fprintf(LOGFILE, "Relocating...\n");
	}
	rewind(Rtmp);
#ifdef INCORE
	if (fread(Buffer, 1, n, Objf) != n) readerr();
	for (i=0; i<nrel; i++) {
		t = fgetc(Rtmp);
		a = fgetc(Rtmp);
		a += fgetc(Rtmp) << 8;
		paddr = &Buffer[a];
		k = paddr[0] + (paddr[1]<<8);
		k += (t == SCODE? Codep: Datap);
		paddr[0] = k;
		paddr[1] = k>>8;
		if (debug) {
			fprintf(LOGFILE,
				" %04x%c%s",
				a, (char) t, !((i+1)%10)? "\n": "");
			fflush(stdout);
		}
	}
	if (debug) fputc('\n', LOGFILE);
	if (fwrite(Buffer, 1, n, Ctmp) != n) writeerr();
#else
	rewind(Rbuf);
	for (i=n; i; ) {
		k = (char *) i < (char *) BUFSIZ? i: BUFSIZ;
		if (fread(Buf, 1, k, Objf) != k) readerr();
		if (fwrite(Buf, 1, k, Rbuf) != k) writeerr();
		i -= k;
	}
	for (i=0; i<nrel; i++) {
		t = fgetc(Rtmp);
		a = fgetc(Rtmp);
		a += fgetc(Rtmp) << 8;
		ufseek(Rbuf, a, SEEK_SET);
		k = fgetc(Rbuf);
		k += fgetc(Rbuf)<<8;
		k += (t == SCODE? Codep: Datap);
		ufseek(Rbuf, a, SEEK_SET);
		fputc(k, Rbuf);
		fputc(k>>8, Rbuf);
		if (debug) {
			fprintf(LOGFILE,
				" %04x%c%s",
				a, (char) t, !((i+1)%10)? "\n": "");
			fflush(stdout);
		}
	}
	if (debug) fputc('\n', LOGFILE);
	rewind(Rbuf);
	for (i=n; i; ) {
		k = (char *) i < (char *) BUFSIZ? i: BUFSIZ;
		if (fread(Buf, 1, k, Rbuf) != k) readerr();
		if (fwrite(Buf, 1, k, Ctmp) != k) writeerr();
		i -= k;
	}
#endif
	old = Codep;
	cmp = (Codep += n);
	if ((char *) cmp < (char *) old || (char *) Codep > (char *) CODESZ)
		fatal("code segment overflow");
	for (i=size=Ohd[ODATASZ] + (Ohd[ODATASZ+1]<<8); i; ) {
		if (i >= BUFSIZ || i < 0)
			n = BUFSIZ;
		else
			n = i;
		if (fread(Buf, 1, n, Objf) != n) readerr();
		if (fwrite(Buf, 1, n, Dtmp) != n) writeerr();
		off += n;
		i -= n;
	}
	old = Datap;
	cmp = (Datap += size);
	if ((char *) cmp < (char *) old || (char *) Datap > (char *) DATASZ)
		fatal("data segment overflow");
	if (debug) fprintf(LOGFILE, "%d DATA bytes\n", size);
}

void resolve(int ptr, int addr, int class) {
	int	i, old;
	int	va, la;

	i = 0;
	while (ptr != 0xffff) {
		la = Extsym[ptr+MKADDR] + (Extsym[ptr+MKADDR+1]<<8);
		if (debug)
			fprintf(LOGFILE,
				" %04x%s%s",
				la, (Extsym[ptr+MKFLAG] & MKREL)? "R": "A",
				!(++i%10)? "\n": "");
		ufseek(Ctmp, la, SEEK_SET);
		if (Extsym[ptr+MKFLAG] & MKREL) {
			va = addr-la-2;
			fputc(va & 0xff, Ctmp);
			fputc(va >> 8, Ctmp);
		}
		else {
			va = (fgetc(Ctmp) & 0xff);
			va += (fgetc(Ctmp) << 8);
			va += addr;
			ufseek(Ctmp, la, SEEK_SET);
			fputc(va & 0xff, Ctmp);
			fputc(va >> 8, Ctmp);
		}
		old = ptr;
		ptr = Extsym[ptr+MKPTR] + (Extsym[ptr+MKPTR+1]<<8);
		Extsym[old+MKADDR] = 0;
	}
	fseek(Ctmp, 0, SEEK_END);
	clearerr(Ctmp);
	if (debug) fputc('\n', LOGFILE);
}

void try_reslv(void) {
	int	i, j, ptr, addr;

	for (i=0; i<SYM_SZ; i+=SSIZE) {
		if (Symtab[i+SCLASS] == EXTRN) {
			for (j=0; j<SYM_SZ; j+=SSIZE) {
				if (Symtab[j+SCLASS] == PUBLIC) {
					if (!strcmp(&Symtab[i], &Symtab[j])) {
						ptr = (Symtab[i+SMARK]&0xff)+
						      (Symtab[i+SMARK+1]<<8);
						addr = (Symtab[j+SADDR]&0xff)+
						       (Symtab[j+SADDR+1]<<8);
						if (debug)
							fprintf(LOGFILE,
						"resolving:\t%s (%04x) ...\n",
							&Symtab[i], addr);
						resolve(ptr, addr,
							Symtab[j+SSEGMT]);
						Symtab[i] = 0;
						Symtab[i+SCLASS] = 0;
						break;
					}
				}
			} /* <-- break goes here */
		}
	}
}

void mkxhdr(int valid) {
	int	k;

	rewind(Exef);
	memcpy((char *)&Xhd[X_MAGIC], valid? XMAGIC: "__", 2);
	Xhd[X_NRELOC] = N_dosrel;
	Xhd[X_NRELOC+1] = N_dosrel>>8;
	k = ((XORELOC + (N_dosrel<<2)) >> 4)+1;
	k = (k + 31) / 32 * 32;
	Xhd[X_HDSIZE] = k;
	Xhd[X_HDSIZE+1] = k>>8;
	Xhd[X_MAXAL] = XMAXAL;
	Xhd[X_MAXAL+1] = XMAXAL>>8;
	Xhd[X_STKPTR] = Stkinit;
	Xhd[X_STKPTR+1] = Stkinit>>8;
	Xhd[X_CKSUM] = XCKSUM;
	Xhd[X_CKSUM+1] = XCKSUM>>8;
	Xhd[X_ENTRY] = XENTRY;
	Xhd[X_ENTRY+1] = XENTRY>>8;
	Xhd[X_CSEG] = XCSEG;
	Xhd[X_CSEG+1] = XCSEG>>8;
	Xhd[X_ORELOC] = XORELOC;
	Xhd[X_ORELOC+1] = XORELOC>>8;
	Xhd[X_OVLID] = XOVL;
	Xhd[X_OVLID+1] = XOVL>>8;
	if (fwrite(Xhd, 1, XHDR_SZ, Exef) != XHDR_SZ)
		writeerr();
}

void bind(void) {
	int	i, n, hdsz, seg2;
	int	t, o, k;
	char	*relp, *slp;

	for (i=0; i<SYM_SZ; i+=SSIZE) {
		if (Symtab[i+SCLASS] == EXTRN) {
			slp = &Symtab[i+SMARK];
			if (*slp != 0xff || slp[0] != 0xff)
				error("unresolved symbol", &Symtab[i]);
		}
	}
	while ((Codep++)%16)
		fputc(0, Ctmp);
	for (hdsz=(N_dosrel<<2)+XORELOC; hdsz%512; hdsz++)
		fputc(0, Exef);
	rewind(Ctmp);
	rewind(Dtmp);
	n = fread(Buf, 1, BUFSIZ, Ctmp);
	while (n) {
		fwrite(Buf, 1, n, Exef);
		n = fread(Buf, 1, BUFSIZ, Ctmp);
	}
	n = fread(Buf, 1, BUFSIZ, Dtmp);
	while (n) {
		fwrite(Buf, 1, n, Exef);
		n = fread(Buf, 1, BUFSIZ, Dtmp);
	}
	fclose(Ctmp);
	fclose(Dtmp);
	seg2 = 0;
	o = t = Codep;
	t += Datap;
	if ((char *) o > (char *) t) seg2 = 128;
	t += hdsz;
	if ((char *) o > (char *) t) seg2 = 128;
	if (t & 0x8000) {
		seg2 += 64;
		t &= 0x7fff;
	}
	k = t/512 + seg2 + 1;
	Xhd[X_NPAGES] = k;
	Xhd[X_NPAGES+1] = k>>8;
	k = t%512-1;
	Xhd[X_NBYTES] = k;
	Xhd[X_NBYTES+1] = k>>8;
	Datap = (Datap>>4) & 0xfff;
	Codep = (Codep>>4) & 0xfff;
	Xhd[X_MINAL] = Datap+1;
	Xhd[X_MINAL+1] = Datap+1>>8;
	k = Codep + Datap + 1;
	Xhd[X_SSEG] = k;
	Xhd[X_SSEG+1] = k>>8;
	if (debug) fprintf(LOGFILE, "DGROUP refs:");
	fseek(Exef, XORELOC, SEEK_SET);
	if (N_dosrel > NDOSREL) fatal("DOSRELOC overflow");
	if (fread(DOSRel, 1, N_dosrel<<2, Exef) != N_dosrel<<2) readerr();
	for (i=0, relp = DOSRel; i<N_dosrel; i++) {
		n = relp[0] | (relp[1] << 8);
		n += hdsz;
		if (debug) fprintf(LOGFILE, " %04x", n);
		relp += 4;
		ufseek(Exef, n, SEEK_SET);
		fputc(Codep&0xff, Exef);
		fputc(Codep>>8, Exef);
	}
	if (debug) fputc('\n', LOGFILE);
}

void wr_symtab(void) {
	int	i;

	fseek(Exef, 0, SEEK_END);
	i = S_MAGIC;
	fputc(S_MAGIC, Exef);
	fputc(0, Exef);
	for (i=0; i<SYM_SZ; i += SSIZE)
		if (Symtab[i])
			if (fwrite(&Symtab[i], 1, SSIZE, Exef) != SSIZE)
				writeerr();
}

void usage(void) {
	printf("usage: sld [-dsv] [-o outfile] [-L libdir] [-f listfile] "
		"[file ...]"
		" [-llib ...]\n");
	exit(1);
}

char *islib(char *name) {
	static char	lib[81];
	int		len;

	if ((len = strlen(name)) < 2) return NULL;
	if (!strcmp(&name[len-2], ".a")) return name;
	if (strncmp(name, "-l", 2)) return NULL;
	strcpy(lib, Libdir);
	strcat(lib, "/lib");
	strcat(lib, name+2);
	strcat(lib, ".a");
	return lib;
}

void loadlib(char *name, int verbose) {
	FILE	*libf;
	int	hdlen, ptr, i, k;
	char	*ip;
	int	pos;
	char	sym[SSIZE];
	int	res_done;

	if ((libf = fopen(name, "rb")) == NULL) {
		error("cannot open library", name);
		return;
	}
	if (fread(Arh, 1, ARH_LEN, libf) != ARH_LEN) readerr();
	if (strncmp(&Arh[AR_MAGIC], A_MAGIC, 2)) {
		error("not a library", name);
		return;
	}
	if (strcmp(&Arh[AR_NAME], SYMDEF)) {
		error("library has no symbol index", name);
		return;
	}
	ip = &Arh[AR_SIZE];
	hdlen = (ip[0] + (ip[1]<<8) + 15) & 0xfff0;
	for (res_done = 1; res_done; ) {
		res_done = 0;
		ptr = ARH_LEN;
		for (; ptr < hdlen+ARH_LEN; ptr += SSIZE) {
			ufseek(libf, ptr, SEEK_SET);
			if (fread(sym, 1, SSIZE, libf) != SSIZE) readerr();
			for (i=0; i<SYM_SZ; i+=SSIZE) {
				if (Symtab[i+SCLASS] == EXTRN) {
					if (!strcmp(&Symtab[i], sym)) {
						ip = &sym[SADDR];
						k = ip[0] + (ip[1]<<8);
						pos = (k<<4) + hdlen;
						ufseek(libf, pos, SEEK_SET);
						if (fread(Arh, 1, ARH_LEN,
						    libf) != ARH_LEN)
							readerr();
						if (verbose > 2)
							printf(
							"loading %s from %s\n",
							&Arh[AR_NAME], name);
						if (strncmp(&Arh[AR_MAGIC],
						    A_MAGIC, 2))
							fatal(
							"bad archive member");
						Objf = libf;
						load(&Arh[AR_NAME]);
						try_reslv();
						clearerr(libf);
						res_done = 1;
					}
				}
			}
		}
	}
	fclose(libf);
}

int main(int argc, char **argv) {
	char	*p, *listfile, buf[81];
	FILE	*list;
	int	verbose, strip;
	int	i;

	strcpy(Outfile, "aout.exe");
	Libdir = ".";
	strip = verbose = 0;
	listfile = NULL;
	for (i=1; i < argc && *argv[i] == '-'; i++) {
		p = argv[i];
		p++;
		if (*p == 'l') break;
		while (*p) switch(*p++) {
			case 'd':	debug = 1; break;
			case 'v':	verbose++; break;
			case 's':	strip = 1; break;
			case 'f':	if (*p == 0) {
						if ((p = argv[++i]) == NULL)
							usage();
					}
					listfile = p;
					p += strlen(p);
					break;
			case 'o':	if (*p == 0) {
						if ((p = argv[++i]) == NULL)
							usage();
					}
					strcpy(Outfile, p);
					p += strlen(p);
					break;
			case 'L':	if (*p == 0) {
						if ((p = argv[++i]) == NULL)
							usage();
					}
					Libdir = p;
					p += strlen(p);
					break;
			default:	usage();
		}
	}
	if (i >= argc && !listfile) usage();
	init();
	Errors = 0;
	mkxhdr(0);
	for (; i < argc; i++) {
		p = argv[i];
		if (islib(p)) break;
		if (verbose > 1) printf("loading %s\n", p);
		if ((Objf = fopen(p, "rb")) == NULL) {
			error("no such file", p);
			continue;
		}
		load(p);
		fclose(Objf);
		try_reslv();
	}
	if (listfile) {
		if ((list = fopen(listfile, "r")) == NULL)
			error("cannot open listfile", listfile);
		else {
			fgets(buf, 81, list);
			while (!feof(list)) {
				buf[strlen(buf)-1] = 0;
				if ((p = islib(buf)) != NULL) {
					if (verbose > 1)
						printf("loading library: %s\n",
						p);
					loadlib(p, verbose);
				}
				else {
					if (verbose > 1)
						printf("loading %s\n", buf);
 					if ((Objf = fopen(buf,"rb")) == NULL) {
						error("no such object file",
							buf);
					}
					else {
						load(buf);
						fclose(Objf);
						try_reslv();
					}
				}
				fgets(buf, 81, list);
			}
			fclose(list);
		}
	}
	for (; i < argc; i++) {
		if ((p = islib(argv[i])) == NULL)
			error("bad library", argv[i]);
		if (verbose > 1) printf("loading library: %s\n", p);
		loadlib(p, verbose);
	}
	bind();
	if (!debug) {
		remove(LDCODE);
		remove(LDDATA);
		remove(LDRLOC);
		remove(LDRBUF);
	}
	mkxhdr(1);
	if (!strip) wr_symtab();
	fclose(Exef);
	if (Errors) {
		printf("%d error(s)\n", Errors);
		remove(Outfile);
		return 1;
	}
	if (verbose) {
		Codep <<= 4;
		Datap <<= 4;
		printf("%5d (%#04x) code bytes\n%5d (%#04x) data bytes\n",
			Codep, Codep, Datap, Datap);
	}
	return 0;
}
